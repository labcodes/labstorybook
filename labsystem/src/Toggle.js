import React from "react";
import PropTypes, { oneOf } from "prop-types";
import { isUndefined } from "lodash";

export default class Toggle extends React.Component {
  static propTypes = {
    color: oneOf(["teal", "purple"]),
    disabled: PropTypes.bool,
    defaultValue: PropTypes.bool,
    value: PropTypes.bool,
    name: PropTypes.string.isRequired,
    handleToggle: PropTypes.func,
  };

  static defaultProps = {
    color: "teal",
    disabled: false,
    defaultValue: false,
    value: undefined,
    handleToggle: undefined,
  };

  constructor(props) {
    super(props);
    const { defaultValue, value, name } = props;
    if (!isUndefined(defaultValue) && !isUndefined(value)) {
      // eslint-disable-next-line no-console
      console.warn(
        `You are setting both value and defaultValue for input ${name} at the same time. We always initialize the toggle with defaultValue. Make sure this is the behaviour you want.`
      );
    }
    this.state = {
      localValue: defaultValue,
    };
  }

  handleOnChange = (e) => {
    const { handleToggle } = this.props;
    if (!isUndefined(handleToggle)) {
      handleToggle(e);
    }

    this.setState((state) => ({ localValue: !state.localValue }));
  };

  render() {
    const { color, name, disabled, value } = this.props;
    const { localValue } = this.state;
    const isChecked = !isUndefined(value) ? value : localValue;

    return (
      <label
        className={`
          lab-toggle
          ${color ? `lab-toggle--${color}` : ""}
          ${disabled ? " lab-toggle--disabled" : ""}
        `}
        htmlFor={name}
      >
        <input
          type="checkbox"
          id={name}
          className="lab-toggle__input"
          checked={isChecked}
          onChange={this.handleOnChange}
          {...(disabled ? { disabled } : undefined)}
        />
        <span className="lab-toggle__slider" />
      </label>
    );
  }
}
