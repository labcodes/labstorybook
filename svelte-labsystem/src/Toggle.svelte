<script>
  export let color = undefined;
  export let disabled = undefined;
  export let value = false;
  export let name;
</script>

<label
  class={`lab-toggle${color ? ` lab-toggle--${color}` : ''}`}
  class:lab-toggle--disabled={disabled}
  htmlFor={name}>
  <input
    type="checkbox"
    id={name}
    class="lab-toggle__input"
    on:change
    bind:checked={value}
    {name}
    {disabled} />
  <span class="lab-toggle__slider" />
</label>
