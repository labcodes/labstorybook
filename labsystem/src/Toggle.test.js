/* eslint-disable no-console */
import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";

import Toggle from "./Toggle";

describe("Toggle", () => {
  it("renders with base props", async () => {
    const component = <Toggle name="test-toggle" />;
    expect(mount(component)).toBeTruthy();
    const renderedComponent = renderer.create(component).toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders as expected when passing disabled as true", async () => {
    const renderedComponent = renderer
      .create(<Toggle name="test-toggle" disabled />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders as expected when passing a different color", async () => {
    const renderedComponent = renderer
      .create(<Toggle color="purple" name="test-toggle" disabled />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("initializes toggled if defaultValue is passed", async () => {
    const mountedComponent = mount(<Toggle name="test-toggle" defaultValue />);
    expect(mountedComponent.find({ checked: true })).toHaveLength(1);
  });

  it("initializes untoggled if defaultValue is passed as false", async () => {
    const mountedComponent = mount(
      <Toggle name="test-toggle" defaultValue={false} />
    );
    expect(mountedComponent.find({ checked: false })).toHaveLength(1);
  });

  it("initializes untoggled if defaultValue is not passed", async () => {
    const mountedComponent = mount(<Toggle name="test-toggle" />);
    expect(mountedComponent.find({ checked: false })).toHaveLength(1);
  });

  it("changes state when input changes", async () => {
    const mountedComponent = mount(<Toggle name="test-toggle" />);

    expect(mountedComponent.find({ checked: false })).toHaveLength(1);
    mountedComponent.find("input").at(0).simulate("change");
    expect(mountedComponent.find({ checked: true })).toHaveLength(1);
  });

  it("calls props.handleToggle passing event when input changes", async () => {
    const mockHandleToggle = jest.fn();
    const mountedComponent = mount(
      <Toggle name="test-toggle" handleToggle={mockHandleToggle} />
    );

    expect(mountedComponent.find({ checked: false })).toHaveLength(1);
    expect(mockHandleToggle).not.toBeCalled();

    mountedComponent.find("input").at(0).simulate("change", { test: "event" });

    expect(mountedComponent.find({ checked: true })).toHaveLength(1);
    expect(mockHandleToggle).toBeCalledWith(
      expect.objectContaining({ test: "event" })
    );
  });
});
