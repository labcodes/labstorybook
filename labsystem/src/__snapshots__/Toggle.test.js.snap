// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle renders as expected when passing a different theme 1`] = `
<label
  className="
          toggle
          toggle--purple
           toggle--disabled
        "
  htmlFor="test-toggle"
>
  <input
    checked={false}
    className="toggle__input"
    disabled={true}
    id="test-toggle"
    onChange={[Function]}
    type="checkbox"
  />
  <span
    className="toggle__slider"
  />
</label>
`;

exports[`Toggle renders as expected when passing disabled as true 1`] = `
<label
  className="
          toggle
          
           toggle--disabled
        "
  htmlFor="test-toggle"
>
  <input
    checked={false}
    className="toggle__input"
    disabled={true}
    id="test-toggle"
    onChange={[Function]}
    type="checkbox"
  />
  <span
    className="toggle__slider"
  />
</label>
`;

exports[`Toggle renders with base props 1`] = `
<label
  className="
          toggle
          
          
        "
  htmlFor="test-toggle"
>
  <input
    checked={false}
    className="toggle__input"
    disabled={false}
    id="test-toggle"
    onChange={[Function]}
    type="checkbox"
  />
  <span
    className="toggle__slider"
  />
</label>
`;
